### PR TITLE
Add per-op generators for sequence_mask, embedding_lookup, gather_nd, boolean_mask, extract_patches (wala/ML#449 Tier 8)

### PR DESCRIPTION
**Pull request by khatchad**
## Summary

Closes the regression Copilot flagged on #236 — when those five ops' `read_data` markers are inlined, the generic `ReadDataFallback` dispatch no longer fires. This adds dedicated `TensorGenerator` subclasses so dispatch stays valid both pre- and post-#236.

| Op | Generator | Output dtype | Output shape |
| --- | --- | --- | --- |
| `tf.sequence_mask` | `SequenceMask` | `BOOL` (fixed; mirrors `Argmax`'s int64) | ⊤ |
| `tf.nn.embedding_lookup` | `EmbeddingLookup` | from `params` | ⊤ |
| `tf.gather_nd` | `GatherNd` | from `params` | ⊤ |
| `tf.boolean_mask` | `BooleanMask` | from `tensor` | ⊤ |
| `tf.image.extract_patches` | `ExtractPatches` | from `images` | ⊤ |

All five extend `PassThroughUnaryTensorGenerator` for the dtype-from-input behavior and override `getDefaultShapes` → `null` (⊤). Precise shape inference for these is non-trivial (combining multiple inputs' shapes, runtime mask contents, or padding/stride arithmetic) and is tracked via wala/ML#449's tier roadmap.

The `dtype` / `output_type` keyword overrides for these ops are not yet surfaced through `tensorflow.xml`; refining the generators to honor them is a clean follow-up — same caveat as `Argmax`.

## Test plan

- [x] `python3.10 tf2_test_<op>.py` runs each fixture without assertion failures.
- [x] `mvn test` clean: `TestTensorflow2Model` 643/0/0/0; sibling modules clean.

This PR is independent of #236 (works pre- and post-inlining); they can land in either order.

### CODE REVIEW

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import com.ibm.wala.ipa.callgraph.propagation.PointsToSetVariable;
+import com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Generator for {@code tf.boolean_mask}. Output dtype is inherited from the {@code tensor} input.
+ * Output shape is left at ⊤ for now: the precise shape depends on the runtime contents of the
+ * boolean {@code mask} (the result is a 1-D dense tensor whose length is the number of {@code True}
+ * entries in the mask), which static analysis cannot recover. See wala/ML#449 (Tier 8).
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java]: Good catch — the "1-D dense tensor" claim was wrong. Fixed in commit `0582ac1a`. Updated to: result rank is generally `tensor.rank - mask.rank + 1`; the runtime truthiness of `mask` is what leaves the masked dimension undetermined, not the rank itself.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/EmbeddingLookup.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.ibm.wala.ipa.callgraph.CGNode;
+import com.ibm.wala.ipa.callgraph.propagation.PointsToSetVariable;
+import com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Generator for {@code tf.nn.embedding_lookup}. Output dtype is inherited from the {@code params}
+ * argument (the embedding table). Output shape is left at ⊤ for now: the precise shape is {@code
+ * (*ids.shape, params.shape[-1])} which requires combining two inputs' shapes — wala/ML#449 (Tier
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/EmbeddingLookup.java]: Right — the formula was over-specific. Fixed in commit `0582ac1a`. The Javadoc now uses the correct general form `ids.shape + params.shape[1:]`.

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 3]
```diff
@@ -332,6 +332,10 @@ public class TestTensorflow2Model extends TestPythonMLCallGraphShape {
 
   private static final TensorType TENSOR_INT64_UNKNOWN_SHAPE = new TensorType(INT_64, null);
 
+  private static final TensorType TENSOR_FLOAT32_UNKNOWN_SHAPE = new TensorType(FLOAT_32, null);
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java]: Right, fixed in commit `0582ac1a`. Dropped the duplicate `TENSOR_FLOAT32_UNKNOWN_SHAPE` and switched to the existing `TENSOR_UNKNOWN_SHAPE_FLOAT32`. Also renamed my new BOOL constant to `TENSOR_UNKNOWN_SHAPE_BOOL` to follow the same naming convention. The older `TENSOR_INT64_UNKNOWN_SHAPE` (different ordering) is left untouched here — broader rename for a separate cleanup pass.

**Comment by khatchad** [comment]
> ## [Codecov](https://app.codecov.io/gh/ponder-lab/ML/pull/239?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) Report
> :x: Patch coverage is `85.36585%` with `12 lines` in your changes missing coverage. Please review.
> :white_check_mark: Project coverage is 68.90%. Comparing base ([`32fadd5`](https://app.codecov.io/gh/ponder-lab/ML/commit/32fadd5ed0332751ba233ac8dad0440d509c92fd?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab)) to head ([`0582ac1`](https://app.codecov.io/gh/ponder-lab/ML/commit/0582ac1a89c3896062c0920b0a026272c68af783?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab)).
> 
> | [Files with missing lines](https://app.codecov.io/gh/ponder-lab/ML/pull/239?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) | Patch % | Lines |
> |---|---|---|
> | [...m/ibm/wala/cast/python/ml/client/SequenceMask.java](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FSequenceMask.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L1NlcXVlbmNlTWFzay5qYXZh) | 50.00% | [4 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> | [...om/ibm/wala/cast/python/ml/client/BooleanMask.java](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FBooleanMask.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L0Jvb2xlYW5NYXNrLmphdmE=) | 71.42% | [2 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> | [...bm/wala/cast/python/ml/client/EmbeddingLookup.java](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FEmbeddingLookup.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L0VtYmVkZGluZ0xvb2t1cC5qYXZh) | 71.42% | [2 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> | [...ibm/wala/cast/python/ml/client/ExtractPatches.java](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FExtractPatches.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L0V4dHJhY3RQYXRjaGVzLmphdmE=) | 71.42% | [2 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> | [...e/com/ibm/wala/cast/python/ml/client/GatherNd.java](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FGatherNd.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L0dhdGhlck5kLmphdmE=) | 71.42% | [2 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/239?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> 
> <details><summary>Additional details and impacted files</summary>
> 
> 
> 
> ```diff
> @@             Coverage Diff              @@
> ##             master     #239      +/-   ##
> ============================================
> + Coverage     68.82%   68.90%   +0.07%     
> - Complexity     2217     2241      +24     
> ============================================
>   Files           213      218       +5     
>   Lines         17753    17835      +82     
>   Branches       2939     2944       +5     
> ============================================
> + Hits          12219    12289      +70     
> - Misses         4386     4398      +12     
>   Partials       1148     1148              
> ```
> </details>
> 
> [:umbrella: View full report in Codecov by Sentry](https://app.codecov.io/gh/ponder-lab/ML/pull/239?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab).   
> :loudspeaker: Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab).
> <details><summary> :rocket: New features to boost your workflow: </summary>
> 
> - :snowflake: [Test Analytics](https://docs.codecov.com/docs/test-analytics): Detect flaky tests, report on failures, and find test suite problems.
> </details>

## Audit of the Codecov-Flagged 12 Missing Lines

| File | Patch % | Missing | What it is |
| --- | --- | --- | --- |
| `SequenceMask.java` | 50% | 4 | `CGNode(node)` ctor (2) + `getInputParameterPosition` / `getInputParameterName` (2 — dead because the BOOL-constant override of `getDefaultDTypes` bypasses the parent's input-arg-reading path). |
| `BooleanMask.java` | 71% | 2 | `CGNode(node)` ctor. |
| `EmbeddingLookup.java` | 71% | 2 | `CGNode(node)` ctor. |
| `ExtractPatches.java` | 71% | 2 | `CGNode(node)` ctor. |
| `GatherNd.java` | 71% | 2 | `CGNode(node)` ctor. |

The 12 missing lines decompose into two structural causes:

1. **5 × `CGNode` ctors are dead because of the silent-skip bug in `TensorGenerator.createManualGenerator`** (wala/ML#468). The ctors exist to satisfy the base-class contract — every concrete `TensorGenerator` subclass should be constructible from either a `PointsToSetVariable` or a `CGNode` — but the manual-walker dispatch table has no arms for these ops, so the ctors are never reached. This isn't a coverage gap that more tests would close; the dispatch is the broken thing. Tracked in wala/ML#468 (silent-skip) and wala/ML#469 (unify dispatch so registration happens once).

2. **2 lines on `SequenceMask` are dead by inheritance.** `SequenceMask` extends `PassThroughUnaryTensorGenerator`, which requires `getInputParameterPosition` / `getInputParameterName`. But `SequenceMask` overrides `getDefaultDTypes` to return `EnumSet.of(BOOL)` unconditionally, bypassing the parent's input-arg-reading path. The two getters compile but are never called. Tracked as a small base-class refactor in wala/ML#472 — make the abstract methods concrete with `UNDEFINED_PARAMETER_POSITION` / `null` defaults, mirroring the existing `getShapeParameterPosition` / `getDTypeParameterPosition` pattern on `TensorGenerator`.

None of the 12 lines hide a runtime bug; they all reflect known structural issues already filed.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/SequenceMask.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Generator for {@code tf.sequence_mask}. Output dtype defaults to {@link DType#BOOL} per the
+ * TensorFlow API; the optional {@code dtype} argument that can override it is not yet surfaced
+ * through {@code tensorflow.xml}, so this generator emits {@code BOOL} unconditionally — same
+ * pattern as {@link Argmax}'s fixed {@code INT64}. Output shape is left at ⊤: the precise shape is
+ * {@code (*lengths.shape, maxlen)} which requires combining one input's shape with another input's
+ * runtime value. See wala/ML#449 (Tier 8).
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/SequenceMask.java]: Right --- the wording was inaccurate. Reworded in commit `27087835`: the dtype is exposed in `tensorflow.xml`'s `paramNames` but is not yet honored by the generator. Added a parenthetical contrast for `argmax`'s `output_type`, which (by contrast) really isn't in the XML at all (`paramNames="self input axis name dimension"`).

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 5]
```diff
@@ -4349,6 +4351,63 @@ public void testArgmin()
     test("tf2_test_argmin.py", "f", 1, 1, Map.of(2, Set.of(TENSOR_INT64_UNKNOWN_SHAPE)));
   }
 
+  /**
+   * Generator-dispatch test for {@code tf.sequence_mask}. Output dtype is fixed at {@code bool}
+   * (the TF default; the {@code dtype} arg override isn't surfaced through {@code tensorflow.xml}
+   * yet), shape is ⊤. See {@link com.ibm.wala.cast.python.ml.client.SequenceMask} (wala/ML#449 Tier
+   * 8).
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java]: Same wording fix --- updated the test Javadoc in commit `27087835` to say the dtype is exposed in the XML but not yet honored by the generator (which emits `bool` unconditionally).
